tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,282 @@
+        :type config: dict
+        :param session: boto3 AWS Session
+        :param logger: Logger object to write to
+        """
+        self.config = config
+        self.logger = logger
+        self.session = session
+
+    def get_splunk_payloads(self, msg, msg_timestamp):
+        """
+        Given a raw c7n message dict, generate a list of payloads to send to the
+        Splunk HEC.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :param msg_timestamp: int timestamp (milliseconds) when message was sent
+          to SQS, for use as event timestamp in Splunk.
+        :type msg_timestamp: int
+        :return: list of Splunk payload dicts
+        :rtype: list
+        """
+        payloads = []
+        events = self.get_splunk_events(msg)
+        indices = self._splunk_indices_for_message(msg)
+        for event in events:
+            for index in indices:
+                payloads.append({
+                    'time': msg_timestamp,
+                    'host': 'cloud-custodian',
+                    'source': '%s-cloud-custodian' % event.get(
+                        'account', 'unknown'
+                    ),
+                    'sourcetype': '_json',
+                    'index': index,
+                    'event': event
+                })
+        return payloads
+
+    def get_splunk_events(self, msg):
+        """
+        Given a raw c7n message dict, generate a list of per-resource event
+        dictionaries to send to Splunk as events.
+
+        :param msg: c7n notification message
+        :type msg: dict
+        :return: list of Splunk HEC Event dicts
+        :rtype: list
+        """
+        # if an event is present, add it to the log
+        if msg.get('event', None) is not None:
+            # add user to the log message, at the top level...
+            user = get_aws_username_from_event(self.logger, msg['event'])
+            if user is not None:
+                msg['event_triggering_user'] = user
+        # get a copy of the message with no resources
+        base_log = dict(msg)
+        base_log.pop('resources')
+        base_log = deepcopy(base_log)
+        # if configured, build and add actions list
+        if self.config.get('splunk_actions_list', False):
+            base_log['actions'] = []
+            for a in msg['policy']['actions']:
+                if isinstance(a, type({})):
+                    base_log['actions'].append(a['type'])
+                else:
+                    base_log['actions'].append(a)
+        # generate a separate Splunk message for each resource
+        logs = []
+        for res in msg['resources']:
+            x = dict(base_log)
+            x['resource'] = res
+            # ensure there's one "tags" element, and it's a dict
+            tmp = self.tags_for_resource(res)
+            if 'Tags' in x['resource']:
+                del x['resource']['Tags']
+            x['resource']['tags'] = tmp
```

Comment:
we really need a copy before we modify resources, else its a shared to other transports. the jsonpatch does look like it defaults to a deep copy even in the place of no operations, so it prior to doing the tags mods, doing a self._prune_log_message would effect a copy.

Reply:
Good catch, thanks. I'd overlooked this since my original code did a deep copy of the whole thing. I've added an explicit shallow copy of the resource, rather than relying on an implementation detail of jsonpatch that theoretically could change in a future version.

Reply:
sounds good re explicit shallow copy.  re original the issue was there as well, the copy was on base_log but resources were pulled/mutated from msg.